rln/src/pm_tree_adapter.rs

```diff
@@ -170,23 +174,41 @@ impl ZerokitMerkleTree for PmTree {
     fn set(&mut self, index: usize, leaf: FrOf<Self::Hasher>) -> Result<()> {
         self.tree
             .set(index, leaf)
-            .map_err(|e| Report::msg(e.to_string()))
+            .map_err(|e| Report::msg(e.to_string()))?;
+        self.cached_leaves_indices[index] = 1;
+        Ok(())
     }
 
     fn set_range<I: IntoIterator<Item = FrOf<Self::Hasher>>>(
         &mut self,
         start: usize,
         values: I,
     ) -> Result<()> {
+        let v = values.into_iter().collect::<Vec<_>>();
         self.tree
-            .set_range(start, values)
-            .map_err(|e| Report::msg(e.to_string()))
+            .set_range(start, v.clone().into_iter())
+            .map_err(|e| Report::msg(e.to_string()))?;
+        for i in start..v.len() {
+            self.cached_leaves_indices[i] = 1
+        }
+        Ok(())
     }
 
     fn get(&self, index: usize) -> Result<FrOf<Self::Hasher>> {
         self.tree.get(index).map_err(|e| Report::msg(e.to_string()))
     }
 
+    fn get_empty_leaves_indices(&self) -> Vec<usize> {
+        let next_idx = self.leaves_set();
+        self.cached_leaves_indices
+            .iter()
+            .take(next_idx)
+            .enumerate()
+            .filter(|&(_, &v)| v == 0u8)
+            .map(|(idx, _)| idx)
+            .collect()
```

Comment:
not sure if this is the best way to handle finding zero elements
perhaps we can employ a bloom filter and set values there?
quick napkin math suggests that we may use just ~3.5mb of memory vs 33mb

vec approach: 1048576 elements * 32 = 33mb at the cost of good performance
bloom filter: `m = ceil((n * log(p)) / log(1 / pow(2, log(2))))` where n = 1048576, p = 1000k = 3.59mb

Reply:
hm, I need to think about it 

If I think correctly, it will require rebuilding the filter every time we increase next_index? I'm not sure how much cost filter updating

Reply:
I also wanna check false negative probability